pallets/marketplace/src/lib.rs

```diff
@@ -0,0 +1,486 @@
+	use frame_system::pallet_prelude::OriginFor;
+
+	pub const RESERVE_ID: ReserveIdentifier = ReserveIdentifier::Marketplace;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	#[pallet::getter(fn tokens)]
+	/// Stores token info
+	pub type Tokens<T: Config> =
+		StorageDoubleMap<_, Twox64Concat, T::NftClassId, Twox64Concat, T::NftInstanceId, TokenInfoOf<T>, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn offers)]
+	/// Stores offer info
+	pub type Offers<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		(T::NftClassId, T::NftInstanceId),
+		Twox64Concat,
+		T::AccountId,
+		OfferOf<T>,
+		OptionQuery,
+	>;
+
+	/// Offer existence check by account and token id
+	#[pallet::storage]
+	#[pallet::getter(fn offers_by_maker)]
+	pub type OffersByMaker<T: Config> = StorageNMap<
+		_,
+		(
+			NMapKey<Blake2_128Concat, T::AccountId>,
+			NMapKey<Blake2_128Concat, T::NftClassId>,
+			NMapKey<Blake2_128Concat, T::NftInstanceId>,
+		),
+		(),
+		ValueQuery,
+	>;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_nft::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		type WeightInfo: WeightInfo;
+		#[pallet::constant]
+		type MinimumOfferAmount: Get<BalanceOf<Self>>;
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Pays a price to the current owner
+		/// Transfers NFT ownership to the buyer
+		/// Disables automatic sell of the NFT
+		///
+		/// Parameters:
+		/// - `class_id`: The identifier of a non-fungible token class
+		/// - `instance_id`: The instance identifier of a class
+		#[pallet::weight(<T as Config>::WeightInfo::buy())]
+		#[transactional]
+		pub fn buy(origin: OriginFor<T>, class_id: T::NftClassId, instance_id: T::NftInstanceId) -> DispatchResult {
+			let sender = ensure_signed(origin.clone())?;
+
+			Self::do_buy(sender, class_id, instance_id, false)
+		}
+
+		/// Set trading price and allow sell
+		/// Setting to NULL will disallow auto sell
+		///
+		/// Parameters:
+		/// - `class_id`: The identifier of a non-fungible token class
+		/// - `instance_id`: The instance identifier of a class
+		/// - `new_price`: price the token will be listed for
+		#[pallet::weight(<T as Config>::WeightInfo::set_price())]
+		#[transactional]
+		pub fn set_price(
+			origin: OriginFor<T>,
+			class_id: T::NftClassId,
+			instance_id: T::NftInstanceId,
+			new_price: Option<BalanceOf<T>>,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			ensure!(
+				pallet_uniques::Pallet::<T>::owner(class_id.into(), instance_id.into()) == Some(sender.clone()),
+				Error::<T>::NotTheTokenOwner
+			);
+
+			Tokens::<T>::try_mutate(class_id, instance_id, |maybe_token_info| -> DispatchResult {
+				let token_info = maybe_token_info.as_mut().ok_or(Error::<T>::NotListed)?;
+
+				token_info.price = new_price;
+
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::TokenPriceUpdated(sender, class_id, instance_id, new_price));
+
+			Ok(())
```

Comment:
Can be simplified so you can get rid of one Ok.

```suggestion
			Tokens::<T>::try_mutate(class_id, instance_id, |maybe_token_info| -> DispatchResult {
				let token_info = maybe_token_info.as_mut().ok_or(Error::<T>::NotListed)?;

				token_info.price = new_price;
				
				Self::deposit_event(Event::TokenPriceUpdated(sender, class_id, instance_id, new_price));

				Ok(())
			})
```